Update Key Vault Secrets README (#2047)
Azure/azure-sdk-for-rust: sdk/keyvault/azure_security_keyvault_secrets/README.md

The `DefaultAzureCredential` will automatically pick up on an Azure CLI authentication. Ensure you are logged in with the Azure CLI:

```azurecli
az login
```

Instantiate a `DefaultAzureCredential` to pass to the client. The same instance of a token credential can be used with multiple clients if they will be authenticating with the same identity.

### Set and Get a Secret

```rust no_run
use azure_identity::DefaultAzureCredential;
use azure_security_keyvault_secrets::{
    models::{SecretBundle, SecretSetParameters},
    ResourceExt, SecretClient,
};

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Create a new secret client
    let credential = DefaultAzureCredential::new()?;
    let client = SecretClient::new(
        "https://your-key-vault-name.vault.azure.net/",

>         credential,

Reviewer: ```suggestion
        credential.clone(),
```

### Thread safety

We guarantee that all client instance methods are thread-safe and independent of each other. This ensures that the recommendation of reusing client instances is always safe, even across threads.

## Examples

The following section provides several code snippets using the `SecretClient`, covering some of the most common Azure Key Vault secrets service related tasks:

* [Create a secret](#create-a-secret)
* [Retrieve a secret](#retrieve-a-secret)
* [Update an existing secret](#update-an-existing-secret)
* [Delete a secret](#delete-a-secret)
* [List secrets](#list-secrets)

### Create a secret

`set_secret` creates a Key Vault secret to be stored in the Azure Key Vault. If a secret with the same name already exists, then a new version of the secret is created.

```rust no_run
use azure_identity::DefaultAzureCredential;
use azure_security_keyvault_secrets::{
    models::{SecretBundle, SecretSetParameters},
    ResourceExt, SecretClient,
};

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let credential = DefaultAzureCredential::new()?;
    let client = SecretClient::new(
        "https://your-key-vault-name.vault.azure.net/",

>         credential,

Reviewer: ```suggestion
        credential.clone(),
```

        None,
    )?;

    // Create a new secret using the secret client.
    let secret_set_parameters = SecretSetParameters {
        value: Some("secret-value".into()),
        ..Default::default()
    };

    let secret: SecretBundle = client
        .set_secret(
            "secret-name".into(),
            secret_set_parameters.try_into()?,
            None,
        )
        .await?
        .into_body()
        .await?;

    println!(
        "Secret Name: {:?}, Value: {:?}, Version: {:?}",
        secret.resource_id()?.name,
        secret.value,
        secret.resource_id()?.version
    );

    Ok(())
}
```

### Retrieve a secret

`get_secret` retrieves a secret previously stored in the Azure Key Vault. Setting the `secret-version` to an empty string will return the latest version.

```rust no_run
use azure_identity::DefaultAzureCredential;
use azure_security_keyvault_secrets::{models::SecretBundle, SecretClient};

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let credential = DefaultAzureCredential::new()?;
    let client = SecretClient::new(
        "https://your-key-vault-name.vault.azure.net/",

>         credential,

Reviewer: ```suggestion
        credential.clone(),
```

        None,
    )?;

    // Retrieve a secret using the secret client.
    let secret: SecretBundle = client
        .get_secret("secret-name".into(), "secret-version".into(), None)
        .await?
        .into_body()
        .await?;

    println!("Secret Value: {:?}", secret.value);

    Ok(())
}
```

### Update an existing secret

`update_secret` updates a secret previously stored in the Azure Key Vault. Only the attributes of the secret are updated. To update the value, call `SecretClient::set_secret` on a secret with the same name.

```rust no_run
use azure_identity::DefaultAzureCredential;
use azure_security_keyvault_secrets::{
    models::{SecretAttributes, SecretUpdateParameters},
    SecretClient,
};
use std::collections::HashMap;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let credential = DefaultAzureCredential::new()?;
    let client = SecretClient::new(
        "https://your-key-vault-name.vault.azure.net/",

>         credential,

Reviewer: ```suggestion
        credential.clone(),
```

        None,
    )?;

    // Update a secret using the secret client.
    let secret_update_parameters = SecretUpdateParameters {
        content_type: Some("text/plain".into()),
        secret_attributes: Some(SecretAttributes::default()),
        tags: Some(HashMap::from_iter(vec![(
            "tag-name".into(),
            "tag-value".into(),
        )])),
    };

    client
        .update_secret(
            "secret-name".into(),
            "".into(),
            secret_update_parameters.try_into()?,
            None,
        )
        .await?
        .into_body()
        .await?;

    Ok(())
}
```

### Delete a secret

> `delete_secret` starts a long-running operation to delete a secret previously stored in the Azure Key Vault. You can retrieve the secret immediately without waiting for the operation to complete.

Reviewer: ```suggestion
`delete_secret` will tell Key Vault to delete a secret but it is not deleted immediately. It will not be deleted until the service-configured data retention period - the default is 90 days - or until you call `purge_secret` on the returned `DeletedSecretBundle.id`.
```

```rust no_run
use azure_identity::DefaultAzureCredential;
use azure_security_keyvault_secrets::SecretClient;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let credential = DefaultAzureCredential::new()?;
    let client = SecretClient::new(
        "https://your-key-vault-name.vault.azure.net/",

>         credential,

Reviewer: ```suggestion
        credential.clone(),
```

        None,
    )?;

    // Delete a secret using the secret client.
    client
        .delete_secret("secret-name".into(), None)
        .await?;

    Ok(())
}
```

### List secrets

This example lists all the secrets in the specified Azure Key Vault. The value is not returned when listing all secrets. You will need to call `SecretClient::get_secret` to retrieve the value.

```rust no_run
use azure_identity::DefaultAzureCredential;
use azure_security_keyvault_secrets::{ResourceExt, SecretClient};
use futures::TryStreamExt;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Create a new secret client
    let credential = DefaultAzureCredential::new()?;
    let client = SecretClient::new(
        "https://your-key-vault-name.vault.azure.net/",

>         credential,

Reviewer: ```suggestion
        credential.clone(),
```

        None,
    )?;

    let mut pager = client.get_secrets(None)?.into_stream();
    while let Some(secrets) = pager.try_next().await? {
        let Some(secrets) = secrets.into_body().await?.value else {
            continue;
        };

        for secret in secrets {
            // Get the secret name from the ID.
            let name = secret.resource_id()?.name;
            println!("Found Secret with Name: {}", name);
        }
    }

    Ok(())
}
```

## Troubleshooting

### General

When you interact with the Azure Key Vault secrets client library using the Rust SDK, errors returned by the service correspond to the same HTTP status codes returned for [REST API] requests.

For example, if you try to retrieve a secret that doesn't exist in your Azure Key Vault, a `404` error is returned, indicating `Not Found`.

```rust no_run
use azure_identity::DefaultAzureCredential;
use azure_security_keyvault_secrets::SecretClient;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let credential = DefaultAzureCredential::new()?;
    let client = SecretClient::new(
        "https://ronnieg-keyvault.vault.azure.net/",

>         credential,

Reviewer: ```suggestion
        credential.clone(),
```

```rust no_run
use azure_identity::DefaultAzureCredential;
use azure_security_keyvault_secrets::{
    models::{SecretAttributes, SecretUpdateParameters},
    SecretClient,
};
use std::collections::HashMap;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let credential = DefaultAzureCredential::new()?;
    let client = SecretClient::new(
        "https://your-key-vault-name.vault.azure.net/",
        credential,
        None,
    )?;

    // Update a secret using the secret client.
    let secret_update_parameters = SecretUpdateParameters {
        content_type: Some("text/plain".into()),
        secret_attributes: Some(SecretAttributes::default()),
        tags: Some(HashMap::from_iter(vec![(
            "tag-name".into(),
            "tag-value".into(),
        )])),
    };

    client
        .update_secret(
            "secret-name".into(),
            "".into(),
            secret_update_parameters.try_into()?,
            None,
        )
        .await?
        .into_body()
        .await?;

    Ok(())
}
```

### Delete a secret

> `delete_secret` starts a long-running operation to delete a secret previously stored in the Azure Key Vault. You can retrieve the secret immediately without waiting for the operation to complete. When [soft-delete] is not enabled for the Azure Key Vault, this operation permanently deletes the secret.

Reviewer: 1. Soft delete is enabled by default and cannot be disabled anymore.
2. Mention of an LRO might make people think we have client support but won't.

So maybe instead,

```suggestion
`delete_secret` will tell Key Vault to delete a secret but it is not deleted immediately. It is not deleted until the service-configured data retention period - the default is 90 days - or until you call `purge_secret` on the returned `DeletedSecretBundle.id`.
```

Author: Added suggestion

> # Azure Key Vault Secret client library for Rust

Reviewer: "secret" is not part of the product name and shouldn't be capitalized. I don't do that in .NET either.
```suggestion
# Azure Key Vault secret client library for Rust
```

Reviewer 2: Also marketing uses a plural `Secrets`

Reviewer 2: Should we use `library` or `crate` terminology?

Reviewer: I was told "no" in the past. Across languages we refer to these as "client libraries". It's that way in the general guidelines as well. As I was told, this is a more abstract concept than whatever we call the container in which the implementation is contained.

Author: Changed all instances of "Secrets" to "secret" and made sure it's always plural when referencing the service branding.

Azure Key Vault is a cloud service that provides a secure storage of secrets, such as passwords and database connection strings.

> The Azure Key Vault Secrets client library allows you to securely store and control the access to tokens, passwords, API keys, and other secrets. This library offers operations to create, retrieve, update, delete, purge, backup, restore, and list the secrets and its versions.

Reviewer 2: In rust they are crates.

[Source code] | [Package (crates.io)] | [API reference documentation] | [Product documentation]

## Getting started

### Install the package

> Install the Azure Key Vault Secrets client library for Rust with [Cargo]:

Reviewer: I recommend lower-casing "secrets" everywhere.
```suggestion
Install the Azure Key Vault secrets client library for Rust with [Cargo]:
```
Best to compare with [.NET's](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/keyvault/Azure.Security.KeyVault.Secrets/README.md). That was the rubric when I lead the KV v-team.

Author: Done

```sh
cargo add azure_security_keyvault_secrets
```

### Prerequisites

* An [Azure subscription].
* An existing Azure Key Vault. If you need to create an Azure Key Vault, you can use the Azure Portal or [Azure CLI].
* Authorization to an existing Azure Key Vault using either [RBAC] (recommended) or [access control].

If you use the Azure CLI, replace `<your-resource-group-name>` and `<your-key-vault-name>` with your own, unique names:

```azurecli
az keyvault create --resource-group <your-resource-group-name> --name <your-key-vault-name>
```

### Install dependencies

Add the following crates to your project:

```sh
cargo add azure_identity azure_core serde_json tokio
```

### Authenticate the client

> In order to interact with the Azure Key Vault service, you'll need to create an instance of the `SecretClient` struct. You need a **vault url**, which you may see as "DNS Name" in the portal, and credentials to instantiate a client object.

Reviewer 2: ```suggestion
In order to interact with the Azure Key Vault service, you'll need to create an instance of the `SecretClient`. You need a **vault url**, which you may see as "DNS Name" in the portal, and credentials to instantiate a client object.
```

The example shown below use a `DefaultAzureCredential`, which is appropriate for most scenarios including local development and production environments. Additionally, we recommend using a managed identity for authentication in production environments. You can find more information on different ways of authenticating and their corresponding credential types in the [Azure Identity] documentation.

The `DefaultAzureCredential` will automatically pick up on an Azure CLI authentication. Ensure you are logged in with the Azure CLI:

```azurecli
az login
```

Instantiate a `DefaultAzureCredential` to pass to the client. The same instance of a token credential can be used with multiple clients if they will be authenticating with the same identity.

### Set and Get a Secret

> ```rust

Reviewer: You need to mark these as `no_run` or tests will try to execute them:

````markdown
```rust no_run
// ...
```
````

See the [guidelines](https://azure.github.io/azure-sdk/rust_introduction.html#rust-documentation-readme) for more information.

use azure_core::{RequestContent, Response};
use azure_identity::DefaultAzureCredential;
use azure_security_keyvault_secrets::{
    models::{SecretBundle, SecretSetParameters},
    SecretClient,
};
use serde_json::to_vec;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Create a new secret client
    let credential = DefaultAzureCredential::new()?;
    let client = SecretClient::new(
        "https://your-key-vault-name.vault.azure.net/",
        credential,
        None,
    )?;

    // Create a new secret using the secret client.

>     let mut secret_set_parameters = SecretSetParameters::default();
>     secret_set_parameters.value = Some("secret-value".to_string());

Reviewer: Easier and idiomatic:
```suggestion
    let secret_set_parameters = SecretSetParameters {
        value: Some("secret-value".to_string()),
        ..Default::default()
    };
```

Author: Was being blocked by the `#[non_exhaustive]` on this before but I believe it's been removed now.

Reviewer: I removed it manually. @jhendrixMSFT were you going to try to do this change before we release beta.1?

Reviewer 3: Yeah it's there now in `v0.8.0` of the emitter.

    // Serialize secret_set_parameters to Vec<u8>
    let secret_set_parameters_bytes: Vec<u8> = to_vec(&secret_set_parameters)?;

> 

Reviewer: No necessary. More work.

    let secret: Response<SecretBundle> = client
        .set_secret(
            "secret-name".to_string(),

>             RequestContent::from(secret_set_parameters_bytes),

Reviewer: More work. You also don't need to import `RequestContent`:
```suggestion
            secret_set_parameters.try_into()?,
```

            None,
        )
        .await?;

>     println!("Response Code: {:?}", secret.status());

Reviewer: Why? Not really useful. It either worked and you print the secret or it didn't.

    // Retrieve a secret using the secret client.
    let secret: SecretBundle = client
        .get_secret(
            "secret-name".to_string(),
            "secret-version".to_string(),
            None,
        )
        .await?
        .into_body()
        .await?;
    println!("{:?}", secret.value);

    Ok(())
}
```

## Key concepts

### KeyVaultSecret

> A `KeyVaultSecret` is the fundamental resource within Azure Key Vault. From a developer's perspective, Azure Key Vault APIs accept and return secret values as strings.

Reviewer: We don't have this type currently but should to align with other languages. I thought @RickWinter was going to do some updates in the `client.tsp` to better align with other languages.

Reviewer: It's called `SecretBundle` currently, but may also be `SecretItem` in some cases. Without hand-writing a client, though, we can't do that with a pure emitted client.

### SecretClient

> A `SecretClient` asynchronous operations in the SDK allowing for selection of a client based on an application's use case. Once you've initialized a `SecretClient`, you can interact with secrets in Azure Key Vault.

Reviewer: "A" infers singular but "operations" is plural.
```suggestion
`SecretClient` asynchronous operations in the SDK allowing for selection of a client based on an application's use case. Once you've initialized a `SecretClient`, you can interact with secrets in Azure Key Vault.
```
This also doesn't make sense. It's the only client. To what selection do you refer? Did you start from Python or Java? Again, for KV anyway, I recommend basing it on .NET. I typically wrote the README first and shared it with other languages.

Author: I did take from .NET. I think the wording is weird because .NET has sync and async operations.

* [Create a secret](#create-a-secret)
* [Retrieve a secret](#retrieve-a-secret)
* [Update an existing secret](#update-an-existing-secret)
* [Delete a secret](#delete-a-secret)
* [Delete and purge a secret](#delete-and-purge-a-secret)
* [List Secrets](#list-secrets)

### Create a secret

`set_secret` creates a `KeyVaultSecret` to be stored in the Azure Key Vault. If a secret with the same name already exists, then a new version of the secret is created.

```rust
use azure_core::{RequestContent, Response};
use azure_identity::DefaultAzureCredential;
use azure_security_keyvault_secrets::{
    models::{SecretBundle, SecretSetParameters},
    SecretClient,
};
use serde_json::to_vec;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let credential = DefaultAzureCredential::new()?;
    let client = SecretClient::new(
        "https://your-key-vault-name.vault.azure.net/",
        credential,
        None,
    )?;

    // Create a new secret using the secret client.
    let mut secret_set_parameters = SecretSetParameters::default();
    secret_set_parameters.value = Some("secret-value".to_string());

    // Serialize secret_set_parameters to Vec<u8>

>     let secret_set_parameters_bytes: Vec<u8> = to_vec(&secret_set_parameters)?;

Reviewer: All these are more complicated than necessary. I can fix them up.

    )?;

    // Retrieve a secret using the secret client.
    let secret: SecretBundle = client
        .get_secret(
            "secret-name".to_string(),
            "secret-version".to_string(),
            None,
        )
        .await?
        .into_body()
        .await?;
    println!("{:?}", secret.value);

    Ok(())
}

```

### Update an existing secret

`update_secret_properties` updates a secret previously stored in the Azure Key Vault. Only the attributes of the secret are updated. To update the value, call `SecretClient::set_secret` on a secret with the same name.

```rust
#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let mut secret = client.get_secret("secret-name").await?;
    secret.properties.content_type = Some("text/plain".to_string());
    secret.properties.tags.insert("foo".to_string(), "updated tag".to_string());

    let updated_secret_properties = client.update_secret_properties(secret.properties).await?;
    println!("{}", updated_secret_properties.name);
    println!("{}", updated_secret_properties.version);
    println!("{}", updated_secret_properties.content_type.unwrap());

    Ok(())
}
```

### Delete a secret

> `start_delete_secret` starts a long-running operation to delete a secret previously stored in the Azure Key Vault. You can retrieve the secret immediately without waiting for the operation to complete. When [soft-delete] is not enabled for the Azure Key Vault, this operation permanently deletes the secret.

Author: Is this true? It's copied from .NET and I assumed if the delete secret API requires an LRO this would still be true for the rust crate.

Reviewer 4: The docs look broken to me. I have some back-story on this. DELETE is not spec'd (in swagger and probably not in TypeSpec either) as a proper LRO. I think that when KeyVault was first created many years ago, the LRO pattern didn't exist. However, deleting a secret is not immediate and is not guaranteed to complete with the DELETE operation returns. So, in other SDKs, we manually wrote code to poll continuously GETtng the secret until 404 returns (or something like this - I forget exactly what the heuristic is). 

But, whatever the heuristic, there is a race condition and implied promise to the customer that we can't keep. So, I'd say that we should expose whatever the service contract actually is and not have our SDK simulate magical LRO functionality. whose behavior we can't guarantee. Also, I believe that most customers simply won't care: they will DELETE a secret, and their app will just move on; the secret will be deleted eventually, why does the app have to wait for it to be fully deleted?

And finally, if the service team thinks there is an issue here with their contract or implementation, then the service team should fix it to benefit ALL their customers which includes customers using an SDK as well as customers not using any SDK.

Reviewer: Barring more feedback about this, I agree with @JeffreyRichter. There's no evidence customers really use this. We may be the only ones in our tests, and I don't plan to exercise this code regularly in our tests because it just eats up time. Cleaning up secrets also implies that we can't run integration tests concurrently and we should be better than that: if we are checking for a collection of secrets for e.g., check only for those names and ignore everything else. Use unique names, etc.

            None,
        )
        .await?;
    println!("Response Code: {:?}", response.status());

    Ok(())
}
```

### Delete a secret

`delete_secret` starts a long-running operation to delete a secret previously stored in the Azure Key Vault. You can retrieve the secret immediately without waiting for the operation to complete. When [soft-delete] is not enabled for the Azure Key Vault, this operation permanently deletes the secret.

```rust
use azure_core::Response;
use azure_identity::DefaultAzureCredential;
use azure_security_keyvault_secrets::{models::DeletedSecretBundle, SecretClient};

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let credential = DefaultAzureCredential::new()?;
    let client = SecretClient::new(
        "https://your-key-vault-name.vault.azure.net/",
        credential,
        None,
    )?;

    // Delete a secret using the secret client.
    let response: Response<DeletedSecretBundle> = client
        .delete_secret(
            "secret_name".to_string(),
            None,
        ).await?;
    println!("Response Code: {:?}", response.status());

    Ok(())
}
```

### Delete and purge a secret

> You will need to wait for the long-running operation to complete before trying to purge or recover the secret. You can do this by calling `update_status` in a loop as shown below:

Author: This is taken from the .NET KeyVault Secrets `README`. Since the rust crate's delete method returns a response, I am unsure how we would query if the operation has completed.

Reviewer: All languages return a response. There's just no model. You still `.await` the response.

Author: My question was more on if we need to wait for the (not really an LRO) LRO to finish before the secret can be purged? 
Or if we even do have to wait for that?

Author: @heaths, 
I tested out the delete and purge scenario and I get the following error: 
```
Error: Error { context: Full(Custom { kind: HttpResponse { status: Conflict, error_code: Some("Conflict") }, error: HttpError { status: Conflict, details: ErrorDetails { code: Some("Conflict"), message: Some("Secret is currently being deleted.") }, .. } }, "server returned error status which will not be retried: 409") }
```
Looks like you do need to wait for the delete to complete before running the purge. 
Here's some sample code I wrote to do the polling Jeff mentioned: 
``` rust
use azure_identity::DefaultAzureCredential;
use azure_security_keyvault_secrets::{models::SecretSetParameters, ResourceExt, SecretClient};

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let credential = DefaultAzureCredential::new()?;
    let client = SecretClient::new(
        "https://ronnieg-keyvault.vault.azure.net/",
        credential,
        None,
    )?;

    println!("Setting a secret...");
    // Set a secret using the secret client.
    let set_secret_parameters = SecretSetParameters {
        value: Some("secret-value".into()),
        ..Default::default()
    };

    let secret = client
        .set_secret(
            "secret-name".into(),
            set_secret_parameters.try_into()?,
            None,
        )
        .await?
        .into_body()
        .await?;

    println!("Deleting the secret...");
    // Delete a secret using the secret client.
    client
        .delete_secret(secret.resource_id()?.name, None)
        .await?;

    println!("Attempting to get the secret until it is deleted...");
    loop {
        // Attempt to get secret until it is deleted
        let response = client
            .get_secret(
                secret.resource_id()?.name,
                secret.resource_id()?.version.unwrap_or_default(),
                None,
            )
            .await;

        if response.is_err() {
            break;
        }
    }

    println!("Purging the deleted secret...");
    // Purge deleted secret using the secret client.
    client
        .purge_deleted_secret(secret.resource_id()?.name, None)
        .await?;

    println!("Operation completed successfully.");
    Ok(())
}
```

This worked for me. In the README we can take out the section to set the secret.

Reviewer: That's why we generated pseudo-LROs for other languages. I mean, how bad do we need to document this? Hardly anyone uses it. If you want, I can write up a recorded test for it.

Author: I'm fine removing this example. I think we have enough examples without it. We can wait for user feedback on this, then add documentation on how to handle it.

Reviewer: I'd recommend instead something like,
```suggestion
`purge_secret` will immediately delete a secret using the `DeletedSecretBundle.id` returned from `delete_secret` unless the service is configured with purge protection, which is disabled by default for Key Vault.

You may not be able to delete a secret immediately after calling `delete_secret`, however. Key Vault may take some time to actually delete the secret, so it's best to retry purging in a loop.
```

    let response: Response<DeletedSecretBundle> = client
        .delete_secret("secret_name".to_string(), None)
        .await?;
    println!("Delete Response Code: {:?}", response.status());

    // Purge deleted secret using the secret client.
    let response: Response<()> = client
        .purge_deleted_secret("secret_name".to_string(), None)
        .await?;
    println!("Purge Response Code: {:?}", response.status());

    Ok(())
}

```

### List secrets

This example lists all the secrets in the specified Azure Key Vault. The value is not returned when listing all secrets. You will need to call `SecretClient::get_secret` to retrieve the value.

```rust
use azure_core::Pager;
use azure_identity::DefaultAzureCredential;
use azure_security_keyvault_secrets::{models::SecretListResult, SecretClient};

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let credential = DefaultAzureCredential::new()?;
    let client = SecretClient::new(
        "https://your-key-vault-name.vault.azure.net/",
        credential,
        None,
    )?;

    // List secrets using the secret client.
    let response: Pager<SecretListResult>= client.get_secrets(None)?;

>     println!("{:?}", response);

Author: What is the proper way to work with the Pager struct? I saw it has a stream property but it's private. And no other methods were showing through IntelliSense on the struct.

Author: Resolved with @heaths's example from the tests

### Delete a secret

> `delete_secret` starts a long-running operation to delete a secret previously stored in the Azure Key Vault. You can retrieve the secret immediately without waiting for the operation to complete.

Reviewer: Please use my recommended text. This still doesn't make sense: "You can retrieve the secret immediately without waiting for the operation to complete."

Reviewer: Maybe GitHub is doing something funky here, but it looks like you previously did - though I didn't see it - and now just restored back to what you had.

Author: Hey @heaths, 
I think this comment might have been made against an old commit. I removed that old text and used your first suggestion in e0fb752.

I updated the text again in 411eaab to this new suggestion.
